IntuneAppConfigurationPolicy: Comparison problem in Test-TargetResource

### Description of the issue

I solved an issue there a couple of weeks ago but after PR #4323 from @William-Francillette has been merged this is not working again, but it's worse now since the test actually spits out an error message.

I have a patch which I tested in my tenant and it fixes the problem, PR to be submitted in a few minutes.

### Microsoft 365 DSC Version

1.24.313.1

### Which workloads are affected

other

### The DSC configuration

```powershell
IntuneAppConfigurationPolicy "IntuneAppConfigurationPolicy-IntuneAppConfigurationPolicy_1"
        {
            Assignments           = @(
                MSFT_DeviceManagementConfigurationPolicyAssignments{
                    dataType = '#microsoft.graph.exclusionGroupAssignmentTarget'
                    deviceAndAppManagementAssignmentFilterType = 'none'
                    groupDisplayName = 'DummyGroupExclude'
                    groupId = '053dc89a-be83-411a-bad3-909904b7239e'
                }
                MSFT_DeviceManagementConfigurationPolicyAssignments{
                    dataType = '#microsoft.graph.groupAssignmentTarget'
                    deviceAndAppManagementAssignmentFilterType = 'none'
                    groupDisplayName = 'DummyGroupInclude'
                    groupId = 'b0b8fd3f-af2a-453b-be57-80182d599f02'
                }
            );
            CustomSettings        = @(
                MSFT_IntuneAppConfigurationPolicyCustomSetting{
                    name = 'com.microsoft.intune.mam.managedbrowser.AppProxyRedirection'
                    value = 'false'
                }
                MSFT_IntuneAppConfigurationPolicyCustomSetting{
                    name = 'com.microsoft.intune.mam.managedbrowser.AllowTransitionOnBlock'
                    value = 'true'
                }
                MSFT_IntuneAppConfigurationPolicyCustomSetting{
                    name = 'com.microsoft.outlook.Contacts.LocalSyncEnabled'
                    value = 'true'
                }
                MSFT_IntuneAppConfigurationPolicyCustomSetting{
                    name = 'com.microsoft.outlook.Contacts.LocalSyncEnabled.UserChangeAllowed'
                    value = 'true'
                }
            );
            Credential            = $CredsCredential;
            DisplayName           = "IntuneAppConfigurationPolicy_1";
            Ensure                = "Present";
            Id                    = "44754094-9c69-4ee0-b587-d8af4efd4a7e";
        }
```


### Verbose logs showing the problem

```powershell
VERBOSE: [fv-az618-863]: LCM:  [ *FAILED*Compare  ]     Completed processing compare operation. The operation returned 
False.
[2024-03-14 08:53:57] -  
##[error]Deploy validation failed ❌
##[error]The PowerShell DSC resource C:\Program Files\WindowsPowerShell\Modules\Microsoft365DSC\1.24.313.1\DscResources\MSFT_IntuneAppConfigurationPolicy returned results in a format that is not valid. The results from running Test-TargetResource must be the boolean value True or False.
##[error]PowerShell exited with code '1'.
```


### Environment Information + PowerShell Version

```powershell
OsName               : Microsoft Windows 11 Enterprise                                                                                                                                                             OsOperatingSystemSKU : EnterpriseEdition                                                                                                                                                                           OsArchitecture       : 64-bit                                                                                                                                                                                      WindowsVersion       : 2009                                                                                                                                                                                        WindowsBuildLabEx    : 22621.1.amd64fre.ni_release.220506-1250                                                                                                                                                     OsLanguage           : en-US                                                                                                                                                                                       OsMuiLanguages       : {en-US, en-GB}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                                                                        Name                           Value                                                                                                                                                                               ----                           -----                                                                                                                                                                               PSVersion                      5.1.22621.1778                                                                                                                                                                      PSEdition                      Desktop                                                                                                                                                                             PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                             BuildVersion                   10.0.22621.1778                                                                                                                                                                     CLRVersion                     4.0.30319.42000                                                                                                                                                                     WSManStackVersion              3.0                                                                                                                                                                                 PSRemotingProtocolVersion      2.3                                                                                                                                                                                 SerializationVersion           1.1.0.1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,9 @@
 
 # UNRELEASED
 
+* IntuneAppConfigurationPolicy
+  * Fix comparison in Test-TargetResource
+    FIXES [#4451](https://github.com/microsoft/Microsoft365DSC/issues/4451)
 * DEPENDENCIES
   * Updated DSCParser to version 2.0.0.0.
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -91,7 +91,7 @@
     Get-TargetResource
   * Fixed an issue with the parameter InterfaceTypes from firewallrules defined
     as a string instead of string[]
-* IntuneDeviceConfigurationPKCSCertificatePolicyWindows10
+* IntuneDeviceConfigurationSCEPCertificatePolicyWindows10
   * Add property RootCertificateDisplayName in order to support assigning root
     certificates by display name since their Ids in a blueprint might be from a
     different source tenant

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1
@@ -365,48 +365,40 @@ function Test-TargetResource
     Write-Verbose -Message "Current Values: $(Convert-M365DscHashtableToString -Hashtable $CurrentValues)"
     Write-Verbose -Message "Target Values: $(Convert-M365DscHashtableToString -Hashtable $PSBoundParameters)"
 
-    $testResult = $true
     if ($CurrentValues.Ensure -ne $PSBoundParameters.Ensure)
     {
-        $testResult = $false
+        Write-Verbose -Message "Test-TargetResource returned $false"
+        return $false
     }
+    $testResult = $true
 
-    $ValuesToCheck = $PSBoundParameters
-    $ValuesToCheck.Remove('Credential') | Out-Null
-    $ValuesToCheck.Remove('ApplicationId') | Out-Null
-    $ValuesToCheck.Remove('TenantId') | Out-Null
-    $ValuesToCheck.Remove('ApplicationSecret') | Out-Null
-    $ValuesToCheck.Remove('CustomSettings') | Out-Null
-
-    #region CustomSettings
-    if ($testResult)
+    #Compare Cim instances
+    foreach ($key in $PSBoundParameters.Keys)
     {
-        $source = $PSBoundParameters.CustomSettings
-        $target = $CurrentValues.CustomSettings
-
-        $source = Get-M365DSCDRGComplexTypeToHashtable -ComplexObject $source
-        $testResult = Compare-M365DSCComplexObject `
-            -Source ($source) `
-            -Target ($target)
-
-        if (-Not $testResult)
+        $source = $PSBoundParameters.$key
+        $target = $CurrentValues.$key
+        if ($source.getType().Name -like '*CimInstance*')
         {
-            $testResult = $false
-            break
-        }
-        $ValuesToCheck.Remove('CustomSettings') | Out-Null
-    }
-    #endregion
+            $source = Get-M365DSCDRGComplexTypeToHashtable -ComplexObject $source
 
-    #region Assignments
-    if ($testResult)
-    {
-        $source = Get-M365DSCDRGComplexTypeToHashtable -ComplexObject $PSBoundParameters.Assignments
-        $target = $CurrentValues.Assignments
-        $testResult = Compare-M365DSCIntunePolicyAssignment -Source $source -Target $target
-        $ValuesToCheck.Remove('Assignments') | Out-Null
+            $testResult = Compare-M365DSCComplexObject `
+                -Source ($source) `
+                -Target ($target)
+
+            if ($key -eq 'Assignments')
+            {
+                $testResult = Compare-M365DSCIntunePolicyAssignment -Source $source -Target $target
+            }
+
+            if (-Not $testResult)
+            {
+                $testResult = $false
+                break
+            }
+
+            $ValuesToCheck.Remove($key) | Out-Null
+        }
     }
-    #endregion
 
     if ($testResult)
     {
@@ -415,6 +407,7 @@ function Test-TargetResource
             -DesiredValues $PSBoundParameters `
             -ValuesToCheck $ValuesToCheck.Keys
     }
+
     Write-Verbose -Message "Test-TargetResource returned $TestResult"
 
     return $TestResult

